vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+    /* Get a list of edge ids, which are outgoing edges from a node */
+    edge_range node_out_edges(const RRNodeId& node) const;
+
+    /* Edge-related attributes 
+     * An example to explain the terminology used in RRGraph
+     *          edgeA
+     *  nodeA --------> nodeB
+     *        | edgeB
+     *        +-------> nodeC
+     *
+     *  +----------+----------------+----------------+
+     *  |  Edge Id |  edge_src_node | edge_sink_node |
+     *  +----------+----------------+----------------+
+     *  |   edgeA  |     nodeA      |      nodeB     |
+     *  +----------+----------------+----------------+
+     *  |   edgeB  |     nodeA      |      nodeC     |
+     *  +----------+----------------+----------------+
+     *
+     */
+    size_t edge_index(const RREdgeId& edge) const; /* TODO: deprecate this accessor as outside functions should use RREdgeId */
+    /* Get the source node which drives a edge */
+    RRNodeId edge_src_node(const RREdgeId& edge) const;
+    /* Get the sink node which a edge ends to */
+    RRNodeId edge_sink_node(const RREdgeId& edge) const;
+    /* Get the switch id which a edge represents
+     * using switch id, timing and other information can be found
+     * for any node-to-node connection
+     */
+    RRSwitchId edge_switch(const RREdgeId& edge) const;
+    /* Judge if a edge is configurable or not. 
+     * A configurable edge is controlled by a programmable memory bit
+     * while a non-configurable edge is typically a hard-wired connection
+     * FIXME: need a double-check for the definition
+     */
+    bool edge_is_configurable(const RREdgeId& edge) const;
+    bool edge_is_non_configurable(const RREdgeId& edge) const;
+
+    /* Switch Info */
+    size_t switch_index(const RRSwitchId& switch_id) const; /* TODO: deprecate this accessor as outside functions should use RRSwitchId */
+    /* Get the switch info of a switch used in this RRGraph */
+    const t_rr_switch_inf& get_switch(const RRSwitchId& switch_id) const;
+
+    /* Segment Info */
+    size_t segment_index(const RRSegmentId& segment_id) const; /* TODO: deprecate this accessor as outside functions should use RRSegmentId */
+    /* Get the segment info of a routing segment used in this RRGraph */
+    const t_segment_inf& get_segment(const RRSegmentId& segment_id) const;
+
+    /* Utilities */
+    /* Find the edges connecting two nodes */
+    std::vector<RREdgeId> find_edges(const RRNodeId& src_node, const RRNodeId& sink_node) const;
+    /* Find a node with given features from internal fast look-up */
+    RRNodeId find_node(const short& x, const short& y, const t_rr_type& type, const int& ptc, const e_side& side = NUM_SIDES) const;
+    /* Find the number of routing tracks in a routing channel with a given coordinate */
+    short chan_num_tracks(const short& x, const short& y, const t_rr_type& type) const;
+
+    /* Check if the graph contains invalid nodes/edges etc. */
+    bool is_dirty() const;
```

Comment:
Can we explain what is_dirty() means and how it would arise? I believe it is when you remove an edge or a node, but that should be stated. What can I not do when I am in the is_dirty() state?  How do I get out of it?  There are some comments in netlist.h that cover this idea and should be adapted.